nixosTests.sourcehut: implement proper integration testing

nessdoor opened the pull request:
## Background
During the review process of #245394, it turned out that large parts of the Sourcehut module have been left to rot due to the (apparently) small size of both its user base and the subset of functionalities that most users actually deploy, in addition to the relatively unstable nature of the Sourcehut software, which can take a considerable amount of maintenance effort in order to be kept on track with upstream changes. 

Excluding a sudden rise in popularity of NixOS-deployed Sourcehut instances, this slow drift into functional decay can only be combated through the effective use of automated integration testing, which can quickly point maintainers at breakages following version bumps in either the Sourcehut package set or of the modules the Sourcehut module depends on.

To make matters worse, the Sourcehut module is very complex to understand and refactor, due to the considerable amount of changes introduced over time and the distributed nature of the services it manages. This, in turn, leads to a very challenging review process, which can stall critical PRs for a very long time and quickly exhaust the energy of reviewers.

Proper integration testing is critical to lower this maintenance and reviewing burden, as it enables quick feedback even to large multi-module refactoring efforts and gives reviewers some confidence that the changes being introduced do not break any existing functionality.

## Description of changes
The current integration tests targeting the Sourcehut module are very limited, as they only verify that services come up at boot and that web servers responds to `GET` requests. Not only that, they also just verify three services (`meta`, `git` and part of `builds`), while the list of Sourcehut services is far longer than that.

This PR introduces the basis on which to build better integration tests for the entire module, based on both scripted interaction with the Web front-end and direct querying of the service APIs, as anticipated in #267538. The current objective is to fully test the interaction of the two core components that most people seem to be using: `meta` and `git`. Nonetheless, we also aim at laying the infrastructure necessary to extend the battery of tests, as more sub-modules get updated and refactored over time. The final goal is that of achieving very high inter-service integration testing coverage, complete with webhooks and other notification mechanisms, through a series of subsequent contributions.

### Draft
This PR will begin as a draft, as all the necessary work has not been completed, yet. The decision to "go public" before completion is because this is a somewhat large undertaking, and it is, therefore, very important to de-duplicate effort, signal to the community that things are moving and, hopefully, attract some helping hands along the way.

The draft status will be lifted as soon as the stated objective (`meta` and `git` integrated testing) will be reached within acceptable margins.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nessdoor committed:
nixosTests.sourcehut: remove unused VM image specifications

At the time of this commit, sourcehut.builds is broken and its test
infrastructure is partial at best. The specified image only inflated the closure
size without any added benefit.

nessdoor committed:
nixosTests.sourcehut: test user creation and OAuth token generation

nessdoor committed:
nixosTests.sourcehut: listen on port 80, as well

nessdoor committed:
nixosTests.sourcehut: run tests within a production environment

nessdoor committed:
nixosTests.sourcehut: configure Hut for direct API interaction

nessdoor committed:
nixosTests.sourcehut: test pushing Git repositories

nessdoor committed:
nixosTests.sourcehut: test repository download

nessdoor committed:
nixosTests.sourcehut: add myself as maintainer

nessdoor committed:
nixosTests.sourcehut: refactor script to use subtests

nessdoor committed:
nixosTests.sourcehut: factor-out node configuration

nessdoor committed:
nixosTests.sourcehut: split tests belonging to different services

nessdoor commented:
Undrafting this now.

The core integrated functionality of a Git code forge has been tested, and going any deeper than this starts to uncover some non-fatal issues that must be addressed separately (e.g. API errors when querying the GraphQL endpoint directly). Additional tests for the rest of the services will be added once their correct base integration has been verified.

Additionally, the outcome of these tests seem to indicate that #138511 is no longer an issue.

christoph-heiss commented on nixos/tests/sourcehut/git.nix:
```diff
@@ -0,0 +1,98 @@
+    ''
+      start_all()
+      machine.wait_for_unit("multi-user.target")
+      # For some reason, OpenSSH fails to start on boot with code 226/NAMESPACE
+      machine.systemctl("start sshd.service")
```
```suggestion
```

This actually broke the testsuite completely for me, as the failed ssh service caused the test VM to abort. Also, this shouldn't make it into the testsuite anyway, IMO.

It seems to happen if `/var/lib/sourcehut/gitsrht/repos` (or whatever path the user configured through `services.sourcehut.settings."git.sr.ht".repos`) does not exist when ssh is started, thus it cannot be mounted into the service namespace.

This could (or rather: should?) be solved by ensuring the directory exists, by using `systemd.tmpfiles`. I did the changes locally and it fixes the issue, at least I could not reproduce it anymore. The full diff of my changes:

```diff
diff --git a/nixos/modules/services/misc/sourcehut/default.nix b/nixos/modules/services/misc/sourcehut/default.nix
index aa803d3bb693..f002635d0bb3 100644
--- a/nixos/modules/services/misc/sourcehut/default.nix
+++ b/nixos/modules/services/misc/sourcehut/default.nix
@@ -790,13 +790,21 @@ in
         '';
       };
       systemd.tmpfiles.settings."10-sourcehut-gitsrht" = mkIf cfg.git.enable (
-        builtins.listToAttrs (map (name: {
-          name = "/var/log/sourcehut/gitsrht-${name}";
-          value.f = {
-            inherit (cfg.git) user group;
-            mode = "0644";
-          };
-        }) [ "keys" "shell" "update-hook" ])
+        mkMerge [
+          (builtins.listToAttrs (map (name: {
+            name = "/var/log/sourcehut/gitsrht-${name}";
+            value.f = {
+              inherit (cfg.git) user group;
+              mode = "0644";
+            };
+          }) [ "keys" "shell" "update-hook" ]))
+          {
+            ${cfg.settings."git.sr.ht".repos}.d = {
+              inherit (cfg.git) user group;
+              mode = "0644";
+            };
+          }
+        ]
       );
       systemd.services.sshd = {
         preStart = mkIf cfg.hg.enable ''
diff --git a/nixos/tests/sourcehut/git.nix b/nixos/tests/sourcehut/git.nix
index cb3ca57ddf62..ed184d5d5518 100644
--- a/nixos/tests/sourcehut/git.nix
+++ b/nixos/tests/sourcehut/git.nix
@@ -49,8 +49,6 @@ in
     ''
       start_all()
       machine.wait_for_unit("multi-user.target")
-      # For some reason, OpenSSH fails to start on boot with code 226/NAMESPACE
-      machine.systemctl("start sshd.service")
       machine.wait_for_unit("sshd.service")

       with subtest("Check whether meta comes up"):
```

christoph-heiss replied:
> Should we integrate the fix as part of this PR or as a separate one?

To be honest, I'd just integrate it as part of this PR. It uncovered this bug and more-or-less depends on the fix to properly work, so I don't see why not. :+1: 

> That was my original intention, but since you already did the heavylifting...

I just investigated it, since it broke the testsuite on my machine as a whole, i.e. if would not even get as far as the above workaround. Didn't want to step on anyones toe, though.

nessdoor replied:
Oh no no no, you didn't step on anybody's toes, I'm in fact happy to not be the only one working on this thing! I'll integrate your changes immediately, then.

nessdoor committed:
nixos/sourcehut: ensure that the repos directory exists

Fixes a bug where, if the OpenSSH server starts before the first initialization
of the git service, the SSH service fails to find the corresponding bind mount
and terminates with a 226/NAMESPACE error.

christoph-heiss reviewed (approved):
Whole testsuite runs through w/o issues now. :+1:
Diff also LGTM. 

Thanks for all the work!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1403

SuperSandro2000 commented on nixos/tests/sourcehut/srht-gen-oauth-tok.nix:
```diff
@@ -0,0 +1,31 @@
+    pps.XMLLibXML
+  ]);
+in
+stdenv.mkDerivation rec {
```
This should be placed under normal packages where someone else could reuse it, maybe for their own testing suite.

nessdoor committed:
srht-gen-oauth-tok: init at 0.1

Promote the srht-gen-oauth-tok package to a Nixpkgs package.

nessdoor commented:
@SuperSandro2000 I have moved the package as you suggested, thank you for your comments!
Please, let me know if you have anything else you want to see me change.
